posts/2023/10/hpy_meetup.md

```diff
@@ -0,0 +1,135 @@
+
+1. A process for discussions about replacement APIs
+2. Guidelines for [evolution of the C API](https://github.com/capi-workgroup/api-evolution).
+
+## 2. Petr's Vision for the C API
+
+Petr presented his understanding of how the C API should look in future. He made
+seven suggestions:
+
+1. _Split_ API from ABI so you can compile the API that it would use different ABI
+   variants.
+
+2. _Consistency_: The API should have a consistent behavior. E.g. APIs should
+   not return borrowed references and the exception behavior should be clear by the
+   naming convention.
+
+3. _Versioning_: Essentially, a C extension should include the ABI version it
+   was compiled with.
+
+4. _Errors_: All API functions must be able to signal an error that can be
+   handled.
+
+5. _Concrete types_: The API should provide dedicated C types for certain
+   built-in objects to have type safety already at the C language level.
+
+6. _Blueprint structures_: Creating types and objects should rather be done by
+   using specifications than by allocating empty objects and filling them with
+   content. Just think of `PyType_Spec` or `PyModuleDef`.
+
+7. _Context argument_: Every API function should have a context argument.
+
+Except for points 4 and 5, these goals are very close to HPy's. We, the
+HPy team, have to say that this sounds very good and we would be happy to see
+that in CPython.
+
+## 3. HPy's Lessons Learned
+
+We were also asked to give a talk and to concentrate in the talk on two things:
+
+1. How do handles affect the API, and
+2. What is the reason for the HPy context to contain the function table.
+
+We didn't _advertise_ HPy in our talk but we talked about why opaque handles
+are not only useful but sometimes necessary and how the API would change
+compared to having only `PyObject *`.
+We further showed that having the function table in the context gives us a lot
+of flexibility and finally enables us to switch on the debug mode essentially at
+run time (when an extension is loaded).
+
+We are not sure if we managed to convince everyone that those concepts are useful.
+The biggest problem is that introducing handles and a context argument will
+radically change the API and break a lot of extensions (see also
+[C API Workgroup - API Revolution #2](https://github.com/capi-workgroup/api-revolution/issues/2)
+and [C API Workgroup - API Revolution #7](https://github.com/capi-workgroup/api-revolution/issues/7)).
+
+# Discussions
+
+We had a lot of good discussions and probably the most important one was about
+the [PEP 703 (noGIL)](https://peps.python.org/pep-0703/) implementation and the
+new version of ABI (abi4) that is planned to be compatible with both noGIL and
+GIL builds of CPython. In order to achieve that, Sam, the author of PEP
+703, plans to make `PyObject *` opaque (so it will narrow to the semantics of a
+handle) and also, the `Py_INCREF` will use an indirection over a function
```

Comment:
AFAIK, the indirection is only planned for `abi4` builds targetting Python 3.11 and below. As support for old Python versions is dropped, it's expected to go away (be replaced by a regular function call, or whatever new mechanism will be used).
```suggestion
handle) and also, the `Py_INCREF` will, in some cases, use an indirection over a function
```